### ISSUE TITLE
default type for number is i64 not i32 (Chapter 2 Primitives)

### ISSUE THREAD

**Issue by jido** [rust-lang/rust-by-example#832]
In the example, I can assign an i64 value to "mutable" as follows:
```
    let mut mutable = 12;

    mutable = 3000000000000000000i64;
```

However there is a type error if I force the type of mutable to i32:
```
    let mut mutable: i32 = 12; // Mutable `i32`.

    // Error! The type of a variable can't be changed
    mutable = 3000000000000000000i64;
```
This indicates that integers default to i64.

**Comment by steveklabnik** [comment]
This isn't true; "unconstrained" numbers will be u32, but given that you said `300000....i64`, you provide a constraint that makes it `i64`. If you'd have left that off, then the type would be `i32`.

Any idea on how to word this so that it makes more sense?

**Comment by jido** [comment]
I may have misinterpreted the results but I really believe that integers defaults to i64 based on my testing.

I updated the issue description to explain better.

**Comment by steveklabnik** [comment]
Yes, I understand your issue. My comment was trying to explain why your interpretation is incorrect.

**Comment by ubsan** [comment]
@jido 

Note that this also works:

```rust
let mut x = 0;
x = 0u8;
let mut y = 1;
y = 1u16;
// and similarly, with all the integer types
```

**Comment by jido** [comment]
@ubsan @steveklabnik 
does that mean that the integer variable is unconstrained, but an integer literal defaults to i32?

In that case when is the default literal type i32 ever relevant since it is not applied to the variable in case of mut and it is resized accordingly when assigning to a different byte length variable?

`let mut x = 10 // x is an unconstrained integer (not i32)`
`let y: i16 = 5 // 5 is converted to i16 on assignment`

I suggest removing any reference to i32 by default, instead say that the type is unconstrained integer when declared that way.

**Comment by ubsan** [comment]
@jido for example, `println!("{}", 13);` shall make that `13` an `i32`. That is the type of default it is talking about.

**Comment by jido** [comment]
@ubsan so when it is assigned to a non-mutable variable, the type is i32 and when it is assigned to a mutable variable the type is unconstrained integer? Is this observable in any way?

**Comment by jido** [comment]
This answers the "observable" although I don't know if there are practical uses:
```
let x = 30;
println!("x occupies {} bytes", mem::size_of_val(&x));
```

**Comment by mdinger** [comment]
Sounds like you're confusing how rust which is strongly typed (strictly?)
with other languages which have things like 'objects' which are coercible
to pretty much anything.

There is no "I don't know what this type is yet so I'm gonna make it an
object (or generic integer) until someone calls me with some specific
parameters". Your available types are the primitives. One of them will be
chosen for the first usage.

The compiler can also look at the context of the program to see if it can
infer what the type is actually supposed to be. This is why your first
example was i64, because you have it context to infer the specific type.

**Comment by mdinger** [comment]
In other words, the program types are changing because you're changing the program:

```rust
// What is the type? It could be default unless specified further down
let mut mutable = 12;
// Oh, you specified it so that 12 we heard about should be `i64`
mutable = 3000000000000000000i64;
```

```rust
// We'd use default or look ahead for the type but the type is specified
// here so our work is done.
let mut mutable: i32 = 12;
// Oh no, you changed your mind! Why did you tell me it was i32 already?
// This won't work
mutable = 3000000000000000000i64;
```

**Comment by jido** [comment]
Thank you, that is very insightful. I will mull over it and come back with my suggestion for this page.

**Comment by jido** [comment]
How about replacing:

> Variables can always be type annotated. Numbers may additionally be annotated via a suffix or by default. Integers default to i32 and floats to f64.

With the following:

> Variables can always be type annotated. Numbers may additionally be annotated via a suffix. Number variables which are not subjected to type annotation anywhere in their scope assume a type by default, which is i32 for integers and f64 for floats.

**Comment by mdinger** [comment]
This sentiment seems OK to me but it seems verbose for what it adds. Up to @steveklabnik though. Kinda feel like the wording could be refined either way.

**Comment by nickretallack** [comment]
This statement also leaves out literals' affinity to match other types found in the same expression.  Later on, when playing with the code in [Casting](http://rustbyexample.com/cast.html) I came up with [this](https://is.gd/yFqv9y), which confused me.  In the expression `(-1000) % 10u8`, 1000 is defaulting to u8 and then complaining that it can't be negated!

Please mention this feature as well.

**Comment by jido** [comment]
Thanks for your contribution @nickretallack. Casting comes after Primitives, so I think the text should be in Primitives. Maybe I should say "subjected to type annotation or type coercion"?
Can you think of a way to make the statement less verbose and/or easier to parse?

**Comment by nickretallack** [comment]
I think you should just include some examples in the code sample that show when a type will be chosen other than the default.

- When the number is too large to fit into an i32
- When it must match another type in the same expression

Unfortunately, it's not until [Casting/Literals](http://rustbyexample.com/cast/literals.html) that you introduce a way to check what type you got.  Either that should happen earlier, or more of this discussion should be pulled into the Casting chapter.

**Comment by mdinger** [comment]
This discussion appears to me to be working towards overspecifying 
something which doesn't necessarily need a large specification from a 
user perspective.

Are many rules dictating when a certain inference will be made 
necessary? If it infers incorrectly, then don't rely on inference. If it 
can't infer, then don't rely on inference. If it works correctly, then 
that is great.

The main topic which needs to be realized is that the compiler *will* 
look around for context if it is not given. Examples attempt to 
highlight this. Does it actually matter how far away the context comes 
from? Is it worth pointing out it will still try to infer context [1000 
lines 
away](https://gist.github.com/anonymous/71b1bc20af0e46fe688a57ce12f2c739) 
if it is required? If it will look that far away, is it surprising it 
will look nearby as well?

**Comment by nickretallack** [comment]
@mdinger I think it's worth mentioning, at least.  The current [Casting/Inference](http://rustbyexample.com/cast/inference.html) page is really light on examples.

**Comment by mdinger** [comment]
Ah, okay

**Comment by marioidival** [comment]
I'm closed this issue because #920 was merged
